### ISSUE TITLE
BUG: `iinfo` type annotation breaks with `TypeError: type 'iinfo' is not subscriptable`

### ISSUE THREAD

**Issue by MarkusSintonen** [numpy/numpy#27341]
### Describe the issue:

Trying to use following with Mypy `strict` mode gives an Mypy error:
```python
import numpy
numpy.iinfo  # error: Missing type parameters for generic type "iinfo"  [type-arg]
```
But trying to fix the error breaks it with Python error:
```python
from typing import Any
import numpy
numpy.iinfo[Any]  # TypeError: type 'iinfo' is not subscriptable
```

### Reproduce the code example:

```python
import numpy
numpy.iinfo  # error: Missing type parameters for generic type "iinfo"  [type-arg]
from typing import Any
numpy.iinfo[Any]  # TypeError: type 'iinfo' is not subscriptable
```

### Python and NumPy Versions:

2.0.1
3.12.4 (main, Jun 18 2024, 13:47:41) [Clang 15.0.0 (clang-1500.3.9.4)]

**Comment by charris** [comment]
@jorenham Is this fixed?

**Comment by jorenham** [comment]
Both `numpy.iinfo` and `numpy.finfo` are *only* generic while type-checking (i.e. `if typing.TYPE_CHECKING: ...`) at the moment.

I agree that this is annoying behavior. So I think it would be best if at runtime, `iinfo` and `finfo` are made into generic types as well.

But in the meantime, there are two ways that you can work around this:

## Workaround 1: Monkey-patching

This is rather hacky, but I don't think this will cause any issues in the internal numpy machinery:

```pycon
>>> import numpy as np
>>> from types import GenericAlias
>>> np.iinfo.__class_getitem__ = classmethod(GenericAlias)
>>> np.iinfo[np.int8]
numpy.iinfo[numpy.int8]
```

The main advantage of this approach is that you only need to do this once in your codebase.

## Workaround 2: `TYPE_CHECKING`-dependent type-aliases

This is probably the cleanest "solution", although mypy might complain about duplicate definitions (but that should be suppressible), depending on whether you use `TypeAlias` or Python 3.12+ `type` alias syntax.

```python
from typing import TYPE_CHECKING, Any
import numpy as np

if TYPE_CHECKING:
    _iinfo = np.iinfo[np.integer[Any]]
else:
    _iinfo = np.iinfo
```

This way you can use `_iinfo` instead of `np.iinfo`, and avoid any typing or runtime errors.

**Comment by jorenham** [comment]
So @charris, it isn't yet, but I'll be working on a solution

**Comment by MarkusSintonen** [comment]
@jorenham thanks for the workarounds. We have instead used mypy ignores in the failure cases. Would be great to fix it. So it would work nicely like other numpy [typing](https://numpy.org/doc/stable/reference/typing.html)

**Comment by jorenham** [comment]
I have opened a PR with a (double) fix: #27346
